### ISSUE TITLE
Can UnwrapResult prefer payload error?

### ISSUE THREAD

**Issue by stridera** [reduxjs/redux-toolkit#701]
In unwrapResult, we throw returned.error, which for custom handlers is always "Rejected"  Can we throw payload.error if it exists before throwing returned.error?  (Is there a way to overwrite returned.error that I didn't see in rejectWithValue?)

```javascript
export function unwrapResult<R extends ActionTypesWithOptionalErrorAction>(
  returned: R
): PayloadForActionTypesExcludingErrorActions<R> {
  if ('error' in returned) {
    throw returned.error
  }
  return (returned as any).payload
}
```

**Comment by phryneas** [comment]
The way `createAsyncThunk` is written, an action returned by it will always either have a `payload` or an `error`, but never both.

If has a payload, we consider this either the result you wanted in the first place or the result of an error you handled gracefully, which should not `throw` any more as well.

If you want to do something different, it's probably best you write your own version of it - it's just 5 lines after all.

**Comment by stridera** [comment]
Yeah, I already rewrote my own.  Just trying to give back.

In this case, can we have rejectWithValue return the result in error instead of payload since you don't expect both?

**Comment by phryneas** [comment]
I'm not sure what you mean.

**Comment by stridera** [comment]
For example...
```javascript
export const createUser = createAsyncThunk("user/create", async (data, { rejectWithValue }) => {
  try {
    const response = await userService.createUser(data);
    return response.data;
  } catch (err) {
    if (!err.response) throw err;
    return rejectWithValue(err.response.data.error);
  }
});
```

Inside the reducer for createUser.rejected, action will look like this:
```javascript
{
    error: { message: "Rejected" },
    meta: {...},
    payload: { error: { message: "Descriptive message from the server." } }
}
```

You said that you expect `createAsyncThunk` actions to always have either a payload or an error.  To match this, I would think the `rejectWithValue` handler should put the payload in the error object instead of creating both the error and payload.

**Comment by phryneas** [comment]
Waitasec. That's a bug. 
We really never caught this? :cry: 

The intended behaviour would be to resolve the promise with the `action.payload` object there :/

**Comment by phryneas** [comment]
Thing is, changing that behaviour now would probably break a lot for a lot of people, so I believe that's not a good idea as well :/

I'd suggest keeping this behaviour for now and completely rething this for the next major release, what do you think @markerikson ?

**Comment by stridera** [comment]
This would resolve the unwrapResult issue as well.  :)

**Comment by markerikson** [comment]
If it's legitimately a bug, we could probably justify it in a minor release, but yeah, this kinda stretches things a bit.

**Comment by phryneas** [comment]
Well, it was not intended to do what it does now, (we're really missing a test for this one specific thing. argh!) but it has never done something else, so it might really break people's code if we change it now :(

**Comment by stridera** [comment]
There is a test, but I'm guessing it's wrong.  It specifically checks to ensure error is a vanilla "Rejected" with the promise in payload.
https://github.com/reduxjs/redux-toolkit/blob/master/src/createAsyncThunk.test.ts#L204

**Comment by phryneas** [comment]
Well, that part is kind of correct. 

The problem is how `unwrapResult` behaves with it, and that probably was an oversight on my part back then :confused: 

And unfortunately, there is no test for `unwrapResult` in that situation even though it was especially created for that situation :(

**Comment by stridera** [comment]
Ahh, gotcha.  I wonder how many people it would break if `unwrapResult` returns the payload instead of a vanilla 'Rejected' as it does now.  I would doubt anyone is using the results currently since there really isn't anything there.  Anyone just catching it and throwing away the payload would have the same functionality.

**Comment by markerikson** [comment]
Well, I see less than 200 usages so far on Github, for what that's worth:

https://github.com/search?q=reduxjs%2Ftoolkit+unwrapresult&type=Code

Might be possible to skim through a few of those and get a sense of what people are doing so far.

**Comment by markerikson** [comment]
So just to check, is this a bug with `rejectWithValue`, or with `unwrapResult` specifically?

**Comment by stridera** [comment]
Depends on your point of view.  If you expect `rejectWithValue` to set the error key, then it's a problem with `rejectWithValue`.  (This is what I would expect, especially with @phryneas's comment about not wanting to set payload and error together.  However, he also said the test was correct that does just that.)

So if you expect error and payload to be used in conjunction, then `unwrapResult` should throw the payload error if it exists, instead of the root error.   If `rejectWithValue`'s payload ends up as error, then `unwrapResult` should function correctly as is.

**Comment by phryneas** [comment]
Well, that "error" property is set very explicitly in the code and it might be useful if the action is handled in a reducer, so I'd keep it in.

The problem is that `unwrapResult` was written too generic to work with just anything that has a `payload` and a potential `error` - even though it was created just for the `rejectWithValue` use case.

So a correct behaviour in my eyes would probably be
* `unwrapResult` works only with `fulfilled` and `rejected` actions
* in the case of a `rejected` action it could be debated - as this never worked before we can pretty much decide on it now
  * either it should fulfill the promise with a `rejectWithValue` or reject it with an `error` if not used with `rejectWithValue` - this is what I usually had in mind, but I believe it confused people anyways.
  * or it should reject in both cases, either with the rejected value or the error. This one would more closely mimic current behaviour and probably also more what people expect

**Comment by stridera** [comment]
For what it's worth, here is the version I wrote for myself.  (And also my vote.)

```javascript
  const unwrapResult = (response) => {
    if (response.payload.error) throw response.payload.error;
    if (response.error) throw response.error;
    return response.payload;
  };
```

**Comment by phryneas** [comment]
I'd opt for adding `action.meta.rejectedWithValue` (like we already have `action.meta.aborted` and `action.meta.condition`) and doing something like 

```
export function unwrapResult(action) {
  if (!action) { return } //someone actually called it with `undefined` today, we can totally catch that
  if (action.meta && action.meta.rejectedWithValue) { throw action.payload }
  if (action.error) { throw action.error }
  return action.payload
}
```

**Comment by markerikson** [comment]
So, is this sounding more like a bugfix now?

**Comment by phryneas** [comment]
It is a behavioural change, but one that should not affect anyone unless people catched that error and then checked if it's message was `"Rejected"`. So I'd say, not a bugfix release, but a minor release?

I can do a PR for that tomorrow.

**Comment by dylinmaust** [comment]
Any chance the related PR can be merged? Or, failing that, can someone provide guidance on a workaround for TypeScript users? Neither `ActionTypesWithOptionalErrorAction` nor `PayloadForActionTypesExcludingErrorActions` is exported, and they're both part of the `unwrapResult` signature, so I'm not sure how I could provide a custom `unwrapResult` function and maintain proper typing. Thanks!

**Comment by amertak** [comment]
Redefined like this so I can use custom Error object after unwraping the result when rejectWithValue is used.

```typescript
declare type ActionTypesWithOptionalErrorAction =
	| {
			error: any
			payload?: any
	  }
	| {
			error?: never
			payload: any
	  }

declare type PayloadForActionTypesExcludingErrorActions<T> = T extends {
	error: any
}
	? never
	: T extends {
			payload: infer P
	  }
	? P
	: never

export function unwrap<R extends ActionTypesWithOptionalErrorAction>(
	returned: R,
): PayloadForActionTypesExcludingErrorActions<R> {
	if (returned.error) throw returned.payload ?? returned.error
	return returned.payload
}
```

**Comment by markerikson** [comment]
Fixed by #704 .
